### PR TITLE
Move to conda environment.

### LINKED ISSUE #32: Move to conda environment.
For inspiration, one can look into [Jupyter Docker Stacks](https://jupyter-docker-stacks.readthedocs.io/en/latest/index.html#jupyter-docker-stacks) and, more specifically, in the corresponding [Docker file](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile).

### PR DESCRIPTION
**Pull request by yakutovicha**
fixes #32

### CODE REVIEW

**Comment by ltalirz** [comment]
I see you requested a review on this - I guess this isn't ready yet?

**Comment by yakutovicha** [comment]
yes, it in the draft state. I have the impression that you can't really review cause I have `Ready for review` button.

**Comment by ltalirz** [comment]
Sorry, I was a bit unclear: I could review (using the interface) but it looked not ready with the failing tests.
Also I noticed you continue to install things via pip, so I thought that stuff was still missing.

Anyhow, perhaps it's actually a good idea not to mix conda & pip installs and to just use the conda environment for the time being.

**Comment by yakutovicha** [comment]
>  Also I noticed you continue to install things via pip, so I thought that stuff was still missing.

I was wondering if a package is available through pipy (aiidalab), is it possible to install it through conda? Or should one make a separate conda-package for that?

**Comment by yakutovicha** [comment]
@ltalirz do you know now to make this part faster?
```
RUN git clone https://github.com/aiidalab/jupyterlab-theme && \
    cd jupyterlab-theme && \
     npm install && \
     npm run build && \
     npm run build:webpack && \
     npm pack ./ && \ 
     jupyter labextension install *.tgz && \
    cd ..
```

it takes a lot of time to finish

**Comment by ltalirz** [comment]
> I was wondering if a package is available through pipy (aiidalab), is it possible to install it through conda? Or should one make a separate conda-package for that?

We'll need a conda package for this. But we shouldn't try to do this now.

In general, mixing conda & pip installs is perhaps not a very good idea. So, staying with pip for the moment might be fine  (how do the jupyter guys handle this?)

**Comment by ltalirz** [comment]
Re jupyterlab:

I don't know how to make it faster, but we're currently not (yet) using it, correct?
I.e. you can both move it further up in the Dockerfile, and comment it out for the time being.

**Comment by yakutovicha** [comment]
> how do the jupyter guys handle this?

They do install pip: https://github.com/jupyter/docker-stacks/blob/414b5d749704fc5abf15b5703551f0acb18e189a/base-notebook/Dockerfile#L89

But the jupyter-related stuff is installed through conda: https://github.com/jupyter/docker-stacks/blob/414b5d749704fc5abf15b5703551f0acb18e189a/base-notebook/Dockerfile#L109

**Comment by yakutovicha** [comment]
Okay, so looking at what I have now I decided to move conda installation to the base image, namely `aiida-prerequisites`.  There is no point in having multiple python installations. If we decided to go for conda, let it be the only python that we officially provide

**Comment by yakutovicha** [comment]
and here we go: https://github.com/aiidateam/aiida-prerequisites/pull/12

**Comment by yakutovicha** [comment]
To be merged after https://github.com/aiidateam/aiida-core/pull/3745

**Comment by yakutovicha** [comment]
@ltalirz it is ready. Please review.

File: Dockerfile [hunk 1]
```diff
@@ -23,52 +18,47 @@ RUN apt-get update && apt-get install -y  \
...
 chmod -R +r /opt/pseudos/
 
-# install packages that are not in the aiidalab meta package
-# 'fastentrypoints' is to fix problems with aiida-quantumespresso plugin installation
-RUN pip3 install --upgrade         \
-    'fastentrypoints'              \
-    'tornado==5.1.1'               \
+
+# Install Python packages needed for AiiDA lab.
+RUN pip install --upgrade         \
```
Comment (ltalirz) [inline Dockerfile]: This might be why tornado is updated?
```suggestion
RUN pip install          \
```

File: Dockerfile [hunk 2]
```diff
@@ -23,52 +18,47 @@ RUN apt-get update && apt-get install -y  \
...
+# Activate ipython kernel.
+# RUN python3 -m ipykernel install
 
 # Enable nbserverproxy extension.
 RUN jupyter serverextension enable --sys-prefix --py nbserverproxy
-# Enables better integration with jupyterhub.
+
+# Enables better integration with Jupyter Hub.
 # https://jupyterlab.readthedocs.io/en/stable/user/jupyterhub.html#further-integration
+# Quite a slow part.
```
Comment (ltalirz) [inline Dockerfile]: Perhaps write:  takes ~X minutes to install

File: Dockerfile [hunk 3]
```diff
@@ -23,52 +18,47 @@ RUN apt-get update && apt-get install -y  \
...
 RUN jupyter labextension install @jupyterlab/hub-extension
 
+# TODO: delete, when https://github.com/aiidalab/aiidalab-widgets-base/issues/31 is fixed
+# the fileupload extension also needs to be "installed".
+RUN jupyter nbextension install --sys-prefix --py fileupload
+
 # Install jupyterlab theme.
+# Also quite slow.
```
Comment (ltalirz) [inline Dockerfile]: same here

File: Dockerfile [hunk 4]
```diff
@@ -79,22 +69,27 @@ RUN git clone https://github.com/aiidalab/jupyterlab-theme && \
...
 COPY my_init.d/prepare-aiidalab.sh /etc/my_init.d/80_prepare-aiidalab.sh
 
-# Start Jupyter notebook
+# Start Jupyter notebook.
 COPY opt/start-notebook.sh /opt/
 COPY service/jupyter-notebook /etc/service/jupyter-notebook/run
 
+# Install some useful packages that are not available on PyPi
+RUN conda install --yes -c conda-forge rdkit
+RUN conda install --yes -c openbabel openbabel "tornado<5"
```
Comment (ltalirz) [inline Dockerfile]: what happens if you don't put "tornado <5" here? Does it not see the tornado version that is already installed?

Also, when providing "tornado<5", does it install a second version of tornado over the pip one?
Just curious to understand what is going on here...

File: opt/prepare-aiidalab.sh [hunk 5]
```diff
@@ -6,6 +6,18 @@ set -x
 # Environment.
 export SHELL=/bin/bash
 
+# Activating conda
```
Comment (ltalirz) [inline opt/prepare-aiidalab.sh]: is this not taken care of by the aiida-core container?

File: opt/prepare-aiidalab.sh [hunk 6]
```diff
@@ -66,7 +78,7 @@ if [ ! -e /home/$SYSTEM_USER/apps ]; then
    touch /home/$SYSTEM_USER/apps/__init__.py
    git clone https://github.com/aiidalab/aiidalab-home /home/$SYSTEM_USER/apps/home
    cd /home/$SYSTEM_USER/apps/home
-   git checkout aiida_v1.0
+   git checkout develop
```
Comment (ltalirz) [inline opt/prepare-aiidalab.sh]: Could you make a note to change this in the future?

I guess the proper branch for the container will be `master`, once the migration has been completed.

File: opt/start-notebook.sh [hunk 7]
```diff
@@ -9,7 +9,7 @@ export SHELL=/bin/bash
 
 # Enter home folder and start jupyterhub-singleuser.
 cd /home/${SYSTEM_USER}
-/opt/aiidalab-singleuser                                           \
+/opt/conda/bin/python /opt/aiidalab-singleuser                                           \
```
Comment (ltalirz) [inline opt/start-notebook.sh]: is the conda python not the default python executable?

i.e. shouldn't  `#!/usr/bin/env python` resolve to `/opt/conda/bin/python`?

File: opt/prepare-aiidalab.sh (hunk 5 above)
Comment (yakutovicha) [reply opt/prepare-aiidalab.sh]: No, this is not how it works.

Conda puts some activation code in the `.bashrc` which is not always loaded. For example, if I run something as another user from root, e.g. `su -c 'some command' aiida, the `.bashrc` file is not loaded. This is the case here.

File: opt/prepare-aiidalab.sh (hunk 6 above)
Comment (yakutovicha) [reply opt/prepare-aiidalab.sh]: Made an issue for that: https://github.com/aiidalab/aiidalab-docker-stack/issues/38

File: opt/start-notebook.sh (hunk 7 above)
Comment (yakutovicha) [reply opt/start-notebook.sh]: same as [here](https://github.com/aiidalab/aiidalab-docker-stack/pull/36/files#r376408695)

File: opt/prepare-aiidalab.sh (hunk 5 above)
Comment (ltalirz) [reply opt/prepare-aiidalab.sh]: do we have to replicate this or can we just source the `.bashrc` then ?

File: Dockerfile (hunk 2 above)
Comment (yakutovicha) [reply Dockerfile]: done

File: Dockerfile (hunk 3 above)
Comment (yakutovicha) [reply Dockerfile]: done

File: Dockerfile (hunk 4 above)
Comment (yakutovicha) [reply Dockerfile]: > what happens if you don't put "tornado <5" here? Does it not see the tornado version that is already installed?

Just tested: conda will see that currently installed tornado (that is 4.5.3) is still compatible with openbabel dependencies and won't touch tornado. So I will remove "tornado<5" then.

> Also, when providing "tornado<5", does it install a second version of tornado over the pip one?

No, looks like conda and PyPI packages both get installed in the same folder which is `/opt/conda/lib/python3.7/site-packages`. Just tested it with `signac` package:

```
root@cc607f96dcd2:/opt/jupyterlab-theme# pip install signac
Collecting signac
  Downloading signac-1.3.0-py3-none-any.whl (181 kB)
     |████████████████████████████████| 181 kB 2.4 MB/s
Collecting filelock~=3.0
  Downloading filelock-3.0.12-py3-none-any.whl (7.6 kB)
Collecting deprecation>=2
  Downloading deprecation-2.0.7-py2.py3-none-any.whl (11 kB)
Requirement already satisfied: packaging>=15.0 in /opt/conda/lib/python3.7/site-packages (from signac) (19.2)
Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=15.0->signac) (2.4.6)
Requirement already satisfied: six in /opt/conda/lib/python3.7/site-packages (from packaging>=15.0->signac) (1.14.0)
Installing collected packages: filelock, deprecation, signac
Successfully installed deprecation-2.0.7 filelock-3.0.12 signac-1.3.0


root@cc607f96dcd2:/opt/jupyterlab-theme# pip freeze | grep signac
signac==1.3.0

root@cc607f96dcd2:/opt/jupyterlab-theme# conda list | grep signac
signac                    1.3.0                    pypi_0    pypi

In [1]: import signac

In [2]: signac.__file__
Out[2]: '/opt/conda/lib/python3.7/site-packages/signac/__init__.py'

```

File: Dockerfile (hunk 1 above)
Comment (yakutovicha) [reply Dockerfile]: Indeed, removing upgrade fixes a lot of stuff, also the AiiDA version problem that we discussed offline

Comment (yakutovicha) [reply Dockerfile]: I've put it all in one RUN command

**Pull request merged**